### DISCUSSION TITLE
Texture normalize op

### DISCUSSION THREAD

**Discussion by franckaubry** [cables-gl/cables_docs#158] [Feature Requests] [answered]
### An operator that takes a greyscale texture and remaps/normalize it to fill out the full white -> black range.

When working with black and white textures/noises, I have seen a great need to be able to create textures that are predictable in it's 0-1 range.
This would give a great starting point for any further compositing, especially when working with randomness during the lifespan of the patch and would also make it possible to create very predictable vertex displacement results.

I imagine the op to work in a way that if you feed it Perlin noise (f.ex.) it would result in a remapping like this:

![image](https://user-images.githubusercontent.com/97125854/154270151-49d135a5-d80f-494b-be39-ccb99a32f0d9.png)

**Comment by pandrr** [comment, marked as answer]
![Screenshot 2022-02-16 at 14 23 07](https://user-images.githubusercontent.com/98792/154273445-53ff0384-c580-40e4-b1d0-4292e30cc5da.png)

colormap does that for you, using a gradient you have quite some control to map the colors

**Reply by franckaubry** [reply]
Absolutely! That is what I have been using so far. What I am talking about is an automatic way to do this, even through time (imagine animated noise)

I believe your solution could still be a viable workflow IF there would be some way to get data about the lightest/darkest pixels and then use that info to create an automatic setup to do the ColorMapping.

Basically a user would need to create a system that works like this:
- Retrieving the lightness data
- Remap those values to 0 and 1
- Use this to control the GradientTexture

**Reply by pandrr** [reply]
yeah that would be quite expensive

**Reply by franckaubry** [reply]
Sounds like it indeed.

Is retrieving the brightness values and doing an auto remap expensive as well? I guess the brightness info will have to be retrieved from the GPU, processed on the CPU and then writen to a the texture again... Or are there any GPU only strategies for these kind of problems?

**Reply by pandrr** [reply]
there are, but its like 10 times more expensive then just doing level or mapping them by hand

https://www.uni-koblenz.de/~cg/ws0809/Seminar_GPUProgrammierung/Material/09_Algorithms_Histograms/GPUHistogram_Kubias2007.pdf

**Reply by franckaubry** [reply]
Ok. I get it. Thank for the link.
So so far, an utopic feature request 🤔

**Reply by AMajesticSeaFlapFlap** [reply]
for some procedural noise textures you could also use the formula that generates the noise to figure out min/max values and remap based on that

**Comment by pandrr** [comment]
and yeah, the perlin contrast is annoying somehow
